### PR TITLE
Generate heatmap transforms

### LINKED ISSUE #3328: Heatmap generation transforms
**Is your feature request related to a problem? Please describe.**
Heatmap regression is used to address a number of regression problems. In these approaches, the targets are usually converted into heatmaps, which are directly learned using a CNN. 

**Describe the solution you'd like**
Currently, MONAI provides a number of transforms to process/pre-process input images, however a transform for generating heatmaps does not exist. The feature will essentially be a transform that takes as input the original regression labels, their ranges, heatmap size and variances for evaluating Gaussians on a heatmap grid. These will be written into a dictionary item of choice, e.g. key=label

**Describe alternatives you've considered**
There are some existing implementation of this, e.g. https://github.com/christianpayer/MedicalDataAugmentationTool-HeatmapRegression however they don't provide a generalised/reusable solution to generating heatmaps. 

**More context**
The transform signature may look like as follows:
```
class GenerateHeatmap(Transform):
       def __init__(in_key, out_key, sigmas, step, range):
              ....
              ....
       def __call__(data):
              ....
              ....
           return data
```

### PR DESCRIPTION
**Pull request by eclipse0922**
Fixes #3328 .

### Description

A few sentences describing the changes proposed in this pull request.

This pull request introduces `GenerateHeatmap` and `GenerateHeatmapd` transforms for creating Gaussian heatmaps from landmark coordinates.
The input points are currently expected in ZYX order, but this can be changed to support XYZ if preferred. 
The transforms ~support both batched (B, N, D) and~ only non-batched (N, D) inputs. 

Example notebooks are included for demonstration and will be removed before the PR is merged.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by eclipse0922** [comment]
[2d_mdtest.ipynb](https://github.com/user-attachments/files/22452287/2d_mdtest.ipynb)
[3d_heatmap_pyvista.ipynb](https://github.com/user-attachments/files/22452288/3d_heatmap_pyvista.ipynb)

**Comment by eclipse0922** [comment]
can anybod review this please?

**Comment by eclipse0922** [comment]
image


[scene-export.html](https://github.com/user-attachments/files/22998595/scene-export.html)


2D and 3D example

**Comment by eclipse0922** [comment]
Could you please take a look at this PR when you have a moment? I believe the GenerateHeatmap would be a valuable addition. Thanks.

@KumoLiu @ericspod @Nic-Ma

File: monai/transforms/post/array.py [hunk 1]
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
...
+class GenerateHeatmap(Transform):
+    """
+    Generate per-landmark Gaussian heatmaps for 2D or 3D coordinates.
+
+    Notes:
+        - Coordinates are interpreted in voxel units and expected in (Y, X) for 2D or (Z, Y, X) for 3D.
+        - Target spatial_shape is (Y, X) for 2D and (Z, Y, X) for 3D.
+        - Output layout uses channel-first convention with one channel per landmark:
+            - Non-batched points (N, D): (N, Y, X) for 2D or (N, Z, Y, X) for 3D
+            - Batched points (B, N, D): (B, N, Y, X) for 2D or (B, N, Z, Y, X) for 3D
```
Comment (ericspod) [inline monai/transforms/post/array.py]: Generally we don't expect batch dimensions in inputs to transforms. If the batch dimension is present is this assumed to be 0?

File: monai/transforms/post/array.py [hunk 2]
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
...
+        - Each channel corresponds to one landmark.
```
Comment (ericspod) [inline monai/transforms/post/array.py]:
```suggestion
        - Each channel index corresponds to one landmark.
```

File: monai/transforms/post/array.py [hunk 3]
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
...
+
+    Args:
+        sigma: gaussian standard deviation. A single value is broadcast across all spatial dimensions.
+        spatial_shape: optional fallback spatial shape. If ``None`` it must be provided when calling the transform.
+            A single int value will be broadcast to all spatial dimensions.
```
Comment (ericspod) [inline monai/transforms/post/array.py]: This doesn't seem to be allowed by the type of the argument nor make sense if it's ambiguous whether your input has a batch dimension or not.

File: monai/transforms/post/array.py [hunk 4]
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
...
+        self.truncated = float(truncated)
+        self.normalize = normalize
+        self.torch_dtype = get_equivalent_dtype(dtype, torch.Tensor)
+        self.numpy_dtype = get_equivalent_dtype(dtype, np.ndarray)
+        # Validate that dtype is floating-point for meaningful Gaussian values
+        if not self.torch_dtype.is_floating_point:
+            raise ValueError(f"dtype must be a floating-point type, got {self.torch_dtype}")
+        self.spatial_shape = None if spatial_shape is None else tuple(int(s) for s in spatial_shape)
+
+    def __call__(self, points: NdarrayOrTensor, spatial_shape: Sequence[int] | None = None) -> NdarrayOrTensor:
```
Comment (ericspod) [inline monai/transforms/post/array.py:806]: I realise a lot of computation is being done here specific to how you want to handle landmark heatmaps on a per landmark basis, but much of the code for doing the gaussian computation I think could be replaced with usages of [GaussianFilter](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L542). The arguments for this class can be tweaked to get the same numeric result I think, or at least close enough for the purpose of heatmaps.

File: monai/transforms/post/dictionary.py [hunk 5]
```diff
@@ -508,6 +513,175 @@ def __init__(self, keys: KeysCollection, output_key: str | None = None, num_clas
         super().__init__(keys, ensemble, output_key)
 
 
+class GenerateHeatmapd(MapTransform):
+    """
+    Dictionary-based wrapper of :py:class:`monai.transforms.GenerateHeatmap`.
+    Converts landmark coordinates into gaussian heatmaps and optionally copies metadata from a reference image.
+
+    Args:
+        keys: keys of the corresponding items in the dictionary.
```
Comment (ericspod) [inline monai/transforms/post/dictionary.py]: I think it needs to be clearer here that the keys are naming tensors of landmark points. An example in this docstring of using the class would help a lot.

File: monai/transforms/post/dictionary.py [hunk 6]
```diff
@@ -508,6 +513,175 @@ def __init__(self, keys: KeysCollection, output_key: str | None = None, num_clas
...
+        d = dict(data)
+        for key, out_key, ref_key, static_shape in self.key_iterator(
+            d, self.heatmap_keys, self.ref_image_keys, self.static_shapes
+        ):
+            points = d[key]
+            shape = self._determine_shape(points, static_shape, d, ref_key)
+            # The GenerateHeatmap transform will handle type conversion based on input points
+            heatmap = self.generator(points, spatial_shape=shape)
+            # If there's a reference image and we need to match its type/device
+            reference = d.get(ref_key) if ref_key is not None and ref_key in d else None
```
Comment (ericspod) [inline monai/transforms/post/dictionary.py:624]:
```suggestion
            reference = d.get(ref_key, None)
```

I think this is equivalent.

File: monai/transforms/post/dictionary.py [hunk 7]
```diff
@@ -508,6 +513,175 @@ def __init__(self, keys: KeysCollection, output_key: str | None = None, num_clas
...
+    """
+
+    backend = GenerateHeatmap.backend
+
+    # Error messages
+    _ERR_HEATMAP_KEYS_LEN = "heatmap_keys length must match keys length."
+    _ERR_REF_KEYS_LEN = "ref_image_keys length must match keys length when provided."
+    _ERR_SHAPE_LEN = "spatial_shape length must match keys length when providing per-key shapes."
+    _ERR_NO_SHAPE = "Unable to determine spatial shape for GenerateHeatmapd. Provide spatial_shape or ref_image_keys."
+    _ERR_INVALID_POINTS = "landmark arrays must be 2D or 3D with shape (N, D) or (B, N, D)."
```
Comment (ericspod) [inline monai/transforms/post/dictionary.py]:
```suggestion
    _ERR_INVALID_POINTS = "Landmark arrays must be 2D or 3D with shape (N, D) or (B, N, D)."
```

File: monai/transforms/post/array.py [hunk 8]
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
...
+            self._sigma = (float(sigma),)
+        if truncated <= 0:
+            raise ValueError("truncated must be positive.")
+        self.truncated = float(truncated)
+        self.normalize = normalize
+        self.torch_dtype = get_equivalent_dtype(dtype, torch.Tensor)
+        self.numpy_dtype = get_equivalent_dtype(dtype, np.ndarray)
+        # Validate that dtype is floating-point for meaningful Gaussian values
+        if not self.torch_dtype.is_floating_point:
+            raise ValueError(f"dtype must be a floating-point type, got {self.torch_dtype}")
```
Comment (ericspod) [inline monai/transforms/post/array.py]:
```suggestion
            raise ValueError(f"Argument `dtype` must be a floating-point type, got {self.torch_dtype}")
```
I'd suggest making it clear the word "dtype" refers to an argument, this applies elsewhere in error strings.

File: tests/transforms/test_generate_heatmapd.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+TEST_CASES_WITH_REF.append(
+    [
+        "dict_with_ref_3d_torch",
+        torch.tensor([[2.5, 2.5, 3.0], [5.0, 5.0, 4.0]], dtype=torch.float32),
+        {"sigma": 2.0},
+        (2, 8, 8, 8),
+        torch.float32,
+        True,  # uses reference image
+    ]
+)
```
Comment (ericspod) [inline tests/transforms/test_generate_heatmapd.py]:
```suggestion
TEST_CASES_WITH_REF = [
    [
        "dict_with_ref_3d_numpy",
        np.array([[2.5, 2.5, 3.0], [5.0, 5.0, 4.0]], dtype=np.float32),
        {"sigma": 2.0},
        (2, 8, 8, 8),
        torch.float32,
        True,  # uses reference image
    ],
    [
        "dict_with_ref_3d_torch",
        torch.tensor([[2.5, 2.5, 3.0], [5.0, 5.0, 4.0]], dtype=torch.float32),
        {"sigma": 2.0},
        (2, 8, 8, 8),
        torch.float32,
        True,  # uses reference image
    ]
]
```
I think it's fine to define lists directly like this.

File: tests/transforms/test_generate_heatmap.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+TEST_CASES_2D = []
+for idx, p in enumerate(TEST_NDARRAYS):
+    TEST_CASES_2D.append(
+        [
+            f"2d_basic_type{idx}",
+            p(np.array([[4.2, 7.8], [12.3, 3.6]], dtype=np.float32)),
+            {"sigma": 1.5, "spatial_shape": (16, 16)},
+            (2, 16, 16),
+        ]
+    )
```
Comment (ericspod) [inline tests/transforms/test_generate_heatmap.py]:
```suggestion
TEST_CASES_2D = [
    [
        f"2d_basic_type{idx}",
        p(np.array([[4.2, 7.8], [12.3, 3.6]], dtype=np.float32)),
        {"sigma": 1.5, "spatial_shape": (16, 16)},
        (2, 16, 16),
    ]
    for idx, p in enumerate(TEST_NDARRAYS)
]
```
If you wanted I think defining values like this is fine.

File: tests/transforms/test_generate_heatmap.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        TEST_CASES_DEVICE_DTYPE.append(
+            [
+                f"cpu_{str(dtype).replace('torch.', '')}",
+                torch.tensor([[3.0, 4.0, 5.0]], dtype=torch.float32),
+                {"sigma": 1.2, "spatial_shape": (10, 10, 10), "dtype": dtype},
+                (1, 10, 10, 10),
+                dtype,
+                "cpu",
+            ]
+        )
```
Comment (ericspod) [inline tests/transforms/test_generate_heatmap.py]:
```suggestion
test_device = "cuda:0" if torch.cuda.is_available() else "cpu"
test_dtypes = [torch.float32, torch.float64]
if torch.cuda.is_available():
    test_dtypes.append(torch.float16)
    
TEST_CASES_DEVICE_DTYPE = [
    [
        f"cuda_{str(dtype).replace('torch.', '')}",
        torch.tensor([[3.0, 4.0, 5.0]], dtype=torch.float32, device=test_device),
        {"sigma": 1.2, "spatial_shape": (10, 10, 10), "dtype": dtype},
        (1, 10, 10, 10),
        dtype,
        test_device,
    ]
    for dtype in test_dtypes
]
```

**Review by ericspod — CHANGES_REQUESTED** [review]
Hi @eclipse0922 sorry for the delay in reviewing. I think these classes would be a good addition to MONAI, I had a few comments below but I hadn't gone through the code thoroughly quiet yet. I commented below that these classes could use [GaussianFilter](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L542) instead of defining the gaussian computation over again, this would save reduce line count and so complexity, and should be adaptable to how you're computing the heatmaps. Please have a look at that class, or elsewhere in MONAI that gaussian kernels are used. Thanks!

**Comment by eclipse0922** [comment]
> Hi @eclipse0922 sorry for the delay in reviewing. I think these classes would be a good addition to MONAI, I had a few comments below but I hadn't gone through the code thoroughly quiet yet. I commented below that these classes could use [GaussianFilter](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L542) instead of defining the gaussian computation over again, this would save reduce line count and so complexity, and should be adaptable to how you're computing the heatmaps. Please have a look at that class, or elsewhere in MONAI that gaussian kernels are used. Thanks!

@ericspod 

 Thank you for the  detailed review. I'll look into your comments and incorporate the changes, including the suggestion to use GaussianFilter

**Comment by eclipse0922** [comment]
[2d_mdtest.ipynb](https://github.com/user-attachments/files/23337943/2d_mdtest.ipynb)
[3d_heatmap_pyvista.ipynb](https://github.com/user-attachments/files/23337944/3d_heatmap_pyvista.ipynb)
updated notebook

**Comment by eclipse0922** [comment]
Hi @ericspod, just a friendly reminder regarding this PR.

I have addressed your feedback from the last review, specifically refactoring the code to use `GaussianFilter` instead of the custom implementation. I also fixed the issues reported by the AI review bots.

Could you please take a look again when you have a chance? Thanks!

**Review by ericspod — APPROVED** [review]
Hi @eclipse0922 I think looks good now. We can merge this soon. Thanks!
